Add inlay hints support

Kingwl opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #42073 

**Features:**

- [x] Parameter name hint for arguments
- [x] Type hint for parameter
- [x] Type hint for variable
- [x] Type hint for field
- [x] Type hint for return type
- [x] Type hint for call chains
- [x] Value hint for numerical enums
- [x] Control to show parameter name hint for non-literal arguments or not 
- [x] Control to show type hints for require assigned variable.
- [x] Control to show parameter name if arguments is the same as parameter name

**Something need to consider:**

- Truncated: For now. I have truncate the hints with hard code 30.
- Preference:  We have many options to specified the behavior. Currently we pass them by UserPreferences. 
- Complex types: Many types may very very long after serialization (function like, complex object, etc.). Should we handle them?
- ~~Type Parameter: The root cause is we could inference type from `type node` or `type`. Which one should we take?~~ I think `type` is better.

Thanks!

Kingwl commented:
@mjbvz 
Please take a look about protocol or naming. 
Thanks!

Kingwl commented:
![image](https://user-images.githubusercontent.com/6831019/103076552-f344ff00-4608-11eb-9736-006ccf00e596.png)

Basically work like this.

Kingwl commented:
@typescript-bot pack this.

Kingwl commented:
https://github.com/microsoft/TypeScript/pull/42284 Could Improve the hints.

Kingwl commented:
https://github.com/microsoft/vscode/pull/113285 is nearl done.

Kingwl commented:
microsoft/vscode#113285 is merged. Now we can focus on https://github.com/microsoft/vscode/pull/113412.

@sandersn @mjbvz might be insested on this.

Kingwl commented:
Ummmmmmm.... Up.

DanielRosenwasser commented on src/services/inlineHints.ts:
```diff
@@ -0,0 +1,300 @@
+                return;
+            }
+
+            switch(node.kind) {
```
```suggestion
            switch (node.kind) {

```

DanielRosenwasser commented on src/services/inlineHints.ts:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        function visitCallOrNewExpression(expr: CallExpression | NewExpression) {
+            if (!expr.arguments || !expr.arguments.length) {
+                return;
+            }
+
+            const candidates: Signature[] = [];
+            const signature = checker.getResolvedSignatureForSignatureHelp(expr, candidates);
+            if (!signature || !candidates.length) {
+                return;
+            }
+
+            for (let i = 0; i < expr.arguments.length; ++i) {
+                const arg = expr.arguments[i];
+                if (!preferences.includeInlineNonLiteralParameterNameHints && !isHintableExpression(arg)) {
+                    continue;
+                }
+
+                const parameterName = checker.getParameterIdentifierNameAtPosition(signature, i);
+                if (parameterName) {
+                    if (preferences.includeInlineDuplicatedParameterNameHints || !isIdentifier(arg) || arg.text !== parameterName) {
+                        addNameHints(unescapeLeadingUnderscores(parameterName), makeEmptyRange(expr.arguments[i].getStart()));
```
I'd just capture `expr.arguments`.

```suggestion
            const args = expr.arguments;
            if (!args || !args.length) {
                return;
            }

            const candidates: Signature[] = [];
            const signature = checker.getResolvedSignatureForSignatureHelp(expr, candidates);
            if (!signature || !candidates.length) {
                return;
            }

            for (let i = 0; i < args.length; ++i) {
                const arg = args[i];
                if (!preferences.includeInlineNonLiteralParameterNameHints && !isHintableExpression(arg)) {
                    continue;
                }

                const parameterName = checker.getParameterIdentifierNameAtPosition(signature, i);
                if (parameterName) {
                    if (preferences.includeInlineDuplicatedParameterNameHints || !isIdentifier(arg) || arg.text !== parameterName) {
                        addNameHints(unescapeLeadingUnderscores(parameterName), makeEmptyRange(args[i].getStart()));
```

DanielRosenwasser commented on src/services/inlineHints.ts:
```diff
@@ -0,0 +1,300 @@
+
+        const checker = program.getTypeChecker();
+        const result: InlineHint[] = [];
+        const callExpressionHintableCache = new Map<CallExpression, boolean>();
```
Do we expect ES5 environments anymore? If not, we don't necessarily have support for `new Map()`.

One optimization (don't do it yet) might be to cache across program instances. If you scroll your window slightly, you'll end up having to re-populate this every time.

On the other hand, it would be nice if editors knew not to throw away prior inline hints if you've scrolled so the language service won't get asked about them again.

Again, don't do it! It's unclear if this is even a problem yet.

Kingwl commented:
Another problem is should we cache the result of all nodes in the program? Is that a bit overkill?

Jack-Works commented:
I'm curious when will this feature come to the language service protocol?

minestarks commented:
@Jack-Works you can track the status of the LSP proposal at https://github.com/microsoft/language-server-protocol/issues/956 , but just to be clear, TypeScript doesn't implement Language Server Protocol today. Am I understanding your question right?

Jack-Works commented:
Thanks @minestarks that's really helpful

Kingwl commented:
👀 Friendly ping.

Kingwl commented:
fRiEnDlY PiNg. 👀

Kingwl commented:
FrIeNdLy pInG. 👀

xiaoxiangmoe commented:
早该管管了(ﾟДﾟ)ﾉ

Kingwl commented:
It's 4.4 frame now. So up again😉.

Kingwl commented:
Well. It's renamed as `InlayHints` in vscode.

Kingwl commented:
Updated and renamed as `Inlay Hints`. 

VSCode for now:
![image](https://user-images.githubusercontent.com/6831019/120595786-8403ee80-c475-11eb-99fb-36731b09114c.png)

/cc: @ghiscoding 😉

Kingwl commented:
With this feature, we can do some pretty cool things, like:

**Any detector**

![as_any](https://user-images.githubusercontent.com/6831019/120743889-70678f00-c52c-11eb-8379-1fc720a36892.gif)

xiaoxiangmoe commented:
Any detector！Excited！

Kingwl commented:
@typescript-bot pack this.

DanielRosenwasser commented:
* Call chain: no, we don't necessarily want to support this pattern
* Make sure that JS patterns don't look silly
    * Don't do signature help's global match that we do in JS files
        * if you request signature help in `x.foo(/**/)`, and `foo` is `any`, today we search for any function named `foo` and try to match against that. Make sure you don't do that here.
    * What does type hint functionality on this look like?
      ```js
      var x = require("yadda")
      ```
* Parameter Name Hints
    * Definitely on by default in TS
    * Opt-in on JS
* Type Hints
    * Opt-in in TS
    * Opt-in in JS
* Other stuff? Don't provide functionality.
* Fully variadic functions? (e.g. `console.log`)
    * ​Don't want to show anything (seems to be today's behavior)
    * What about other variadic functions? `arg0`, `arg1`, `arg2`...?
        * Unsure. Can possibly iterate on current behavior.
* `=` vs. `:` for parameter hints?
    * Seems like this was still intuitive, even if it's misleading in the syntax. Keep it as `:`.

Kingwl commented:
Okay.
But don't we provide enum value hints？ It's seems useful.

Currently, all behaviors are controled by options who send by editor(vscode).
Which means we may update the default value of option on vscode side.
BTW: I think they will don't open anything by default in a while when this landed.

ghiscoding commented:
Hello, so I thought that I could provide early feedback as a user who is anxiously waiting for this feature... I've seen this feature from WebStorm while watching some video tutorial and I find that it looks so useful and productive. The option I'm most interested in, is definitely the function parameter hints and I would also follow WebStorm on using the `:` separator. I also think that all the defaults that I can see in this [WebStorm inlay hints](https://www.jetbrains.com/help/webstorm/inlay-hints.html) doc are just perfect. I'm hopeful this will ship in this upcoming release... Happy Coding :)

Jack-Works commented:
Does it support inferred generics?


```ts
declare function foo<T extends number>(t: T): T

foo(1) // should be displayed as foo<1>(1)
```

Kingwl commented:
Hey folks.

Thanks for the response. Currently It's hard to preview the feature.

I'll try to finger out how to make it easy.

Kingwl commented:
>Does it support inferred generics?

There's no support for expression for now. (Basically for declaration)

Kingwl commented:
- > Call chain: no, we don't necessarily want to support this pattern 

    - Done

- > Don't do signature help's global match that we do in JS files

    - Nope
    -  ![image](https://user-images.githubusercontent.com/6831019/121490171-5daaf980-ca07-11eb-8f64-d3b870a264d7.png)


- > What does type hint functionality on this look like

    - Here's an option
    - If option turn on
        - ![requireAssignmentInJsOn](https://user-images.githubusercontent.com/6831019/121489372-9ac2bc00-ca06-11eb-95eb-ee1cb03eec41.png)
    - if option turn off 
        - ![requireAssignmentInJsOff](https://user-images.githubusercontent.com/6831019/121489380-9dbdac80-ca06-11eb-9ef1-3c8791c02621.png)

- Parameter Name Hints
    - Controls by editor 
- Type Hints
    -  Controls by editor too
- Other stuff? Don't provide functionality.
    - I'm still strongly recommends the enum member value hints
- Fully variadic functions? 
    - Shows parameter names at first variadic argument
    -  ![image](https://user-images.githubusercontent.com/6831019/121489916-22a8c600-ca07-11eb-9d05-beff49c42ccb.png)

Kingwl commented:
Hi @mjbvz . 

Would you mind tell me can the built-in typescript extension be re-pack and install?

I have tried but it seems not work.

The typescript features are all down.

Jack-Works commented:
Yes, the enum one is very useful. Please preserve them. Does it support calculated bit flags? Like 

```ts
enum T {
    a = 1 << 20
}

Jack-Works commented:
> Fully variadic functions? (e.g. console.log)
> ​Don't want to show anything (seems to be today's behavior)
> What about other variadic functions? arg0, arg1, arg2...?
> Unsure. Can possibly iterate on current behavior.

Maybe this?

```ts
declare function log(...args: any): void

log(1, 2, 3)
// displayed as log(...args: 1, 2, 3)
```

Kingwl commented:
> Maybe this?

Sounds a good idea.

Kingwl commented:
Looks pretty good.

![image](https://user-images.githubusercontent.com/6831019/121500386-cfd40c00-ca10-11eb-9924-1c1a524849bd.png)

DanielRosenwasser commented:
# Variadic Arguments

I think I'm willing to change my mind about variadic args - that doesn't look too bad.

I still don't know how useful it is for `console.log` though, and that was part of the question: does it make sense to provide a name there? I actually kind of like it.

# JavaScript Type Annotations

I think we specifically should not provide an annotation for anything with `require(...)` in the initializer, even when the flag is on. Maybe @sandersn has an easy way to figure out if a variable declaration contains a `require()`.

# Enum Values

I'm kind of split. It's nice to have a reminder in case you're running into a bug where you're treating an enum in a truthy way.

But I'm skeptical about bitflags. Showing the numeric value doesn't give you any obvious information about what something was OR'd and AND'd with.

Jack-Works commented:
## Enum values

Display for auto-increase enum members is definitely useful. About bitflags, maybe it can be displayed like `0b000000001`?

Kingwl commented:
Ahhhh. That's sounds not good enough.
Let's just keep the auto increased enum value hints.

Kingwl commented:
>I think we specifically should not provide an annotation for anything with require(...) in the initializer, even when the flag is on.

I mean there's two flags
- show hints for variable declaration
- show hints for require assignment variable declaration

DanielRosenwasser commented:
Yeah, I don't think that `require()` calls even need that option.

_____

Another related issue is the case where a type might show `import(...)` which is not ideal.

For example, I don't think it's appropriate to turn

```ts
let x = makeSomething();
```

into

```ts
let x: import("./blah/blah/foo").Thing = makeSomething();
```

_______

One other edge case that I've thought of: one of the lint rules that we employ in our compiler is that function calls where you pass in `null`, `undefined`, `true`, or `false` all have to be prefixed with a comment naming the argument:

```diff
- foo(someVariable, false);
+ foo(someVariable, /*validateInputs*/ false);
```

There's now 2 concerns:

1. What does this look like when you have comments/newlines preceding the argument?
2. Does this duplicate the information in a preceding comment?

Kingwl commented:
> Yeah, I don't think that require() calls even need that option.

Okay.

> Another related issue is the case where a type might show import(...) which is not ideal.

I have some idea:

- Don't shows all import type: 
  -  `import("./blah/blah/foo").Thing` to nothing
  -  `import("./blah/blah/foo")` to nothing
- Strip import type if it's left of a qualified name: 
  - `import("./blah/blah/foo").Thing` to `Thing`
  - `import("./blah/blah/foo")` to nothing
 
Which do you think is better?

> What does this look like when you have comments/newlines preceding the argument?

It's not care about the comments now. 
And as you can see it [here](https://github.com/microsoft/TypeScript/pull/42089/files#diff-34f5f57d5cbbea640ecda13bb110839f275d281b6b7d93a9f0a96d69fa0a2699R149). It's get the position by `getStart` rather than `getFullStart`.

So it should be something like:
```ts

call(
  /*arg*/ [|hints here|]arg
)
```
> Does this duplicate the information in a preceding comment?

Yep. It's seems duplicated.
Should we check the preceding comments and don't provide the hints if the comments is equals to the hints?

Kingwl commented:
- Always not provide the hints who point to a module (shows as import type).
- Remove related option.
- Not provide hints if argument's leading comments has contains the parameter name.

Some issues:
- About option name: I'm not a naming expert. Does them seems ok？

- About leading comments: String contains might a little overkill (parameter would be named very short: a, i, v, etc.). Any better way to detect the comment?

Thanks!

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+
+        function leadingCommentsHasParameterName(commentRanges: CommentRange[] | undefined, name: string) {
+            if (!commentRanges?.length) return false;
+            return commentRanges.some(range => stringContains(sourceFileText.substring(range.pos, range.end), name));
```
I don't know if this is exactly the logic I'd recommend. My gut says it's something like the regex

```
^\s*(originalString)\s*:\s*$
```

Kingwl replied:
Not pretty sure but I guess I got your point.

DanielRosenwasser commented on src/harness/client.ts:
```diff
@@ -646,6 +646,22 @@ namespace ts.server {
+            const request = this.processRequest<protocol.ProvideInlayHintsRequest>(CommandNames.ProvideInlayHints, args);
+            const response = this.processResponse<protocol.ProvideInlayHintsResponse>(request);
+
+            return response.body!.map(item => ({ // TODO: GH#18217
```
Does this need the TODO comment?

Kingwl replied:
Seems yes.

>  return response.body!.map<CodeFixAction>(({ fixName, description, changes, commands, fixId, fixAllDescription }) => // TODO: GH#18217

At line 641.

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+            return displayPartsToString(displayParts);
+        }
+
+        function writeNodeInSignleLine(node: Node, writer: DisplayPartsSymbolWriter) {
```
```suggestion
        function writeNodeInSingleLine(node: Node, writer: DisplayPartsSymbolWriter) {
```

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+                const typeNode = checker.typeToTypeNode(type, /*enclosingDeclaration*/ undefined, flags, singleLineWriter);
+                Debug.assertIsDefined(typeNode, "should always get typenode");
+
+                writeNodeInSignleLine(typeNode, singleLineWriter);
```
```suggestion
                writeNodeInSingleLine(typeNode, singleLineWriter);
```

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+        function printTypeInSingleLine(type: Type) {
+            const flags = NodeBuilderFlags.IgnoreErrors | TypeFormatFlags.AllowUniqueESSymbolType | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope;
+            const displayParts = mapToDisplayParts(writer => {
+                const singleLineWriter = createSignleLineWriter(writer);
```
```suggestion
                const singleLineWriter = createSingleLineWriter(writer);
```

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+            return text;
+        }
+
+        function createSignleLineWriter(writer: DisplayPartsSymbolWriter): DisplayPartsSymbolWriter {
```
```suggestion
        function createSingleLineWriter(writer: DisplayPartsSymbolWriter): DisplayPartsSymbolWriter {
```

DanielRosenwasser replied:
Surely there must be something like this already? @weswigham @andrewbranch

Kingwl replied:
It might be `createSingleLineStringWriter` in `src/compiler/utilities.ts`

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+        visitor(file);
+        return result;
+
+        function visitor(node: Node): true | undefined | void {
```
Can you just make it `true | undefined`?

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+            }
+
+            switch (node.kind) {
+
```
```suggestion
```

DanielRosenwasser commented on src/services/inlayHints.ts:
```diff
@@ -0,0 +1,291 @@
+                case SyntaxKind.ArrowFunction:
+                    cancellationToken.throwIfCancellationRequested();
+
+                    if (!textSpanIntersectsWith(span, node.pos, node.getFullWidth())) {
```
I think this code is in the wrong place - you are likely walking more of the tree than you need. Wouldn't it be better to stop the walk at the very top of this function?

Kingwl commented:
Another problem is about the naming of options:

Which one looks better?

- includeInlay<xxxxx>Hints (Current one)
- include<xxxxx>InlayHints
- provideInlay<xxxxx>Hints
- provide<xxxxx>Hints
- others?

Thanks!

andrewbranch reviewed (changes requested):
A few small nits on implementation; overall the code looks good. I’m going to try it out and see how the experience is. I think the names you have in UserPreferences are fine. Are there any other questions you have or things you want someone to look at specifically?

andrewbranch commented on src/compiler/types.ts:
```diff
@@ -8462,6 +8463,14 @@ namespace ts {
         readonly providePrefixAndSuffixTextForRename?: boolean;
         readonly includePackageJsonAutoImports?: "auto" | "on" | "off";
         readonly provideRefactorNotApplicableReason?: boolean;
+        readonly includeInlayParameterNameHints?: boolean;
+        readonly includeInlayNonLiteralParameterNameHints?: boolean;
+        readonly includeInlayDuplicatedParameterNameHints?: boolean;
```
What are “duplicated” parameter name hints? Do we need all three of these options? What happens if I have `includeInlayParameterNameHints` disabled but the other two enabled? (Sorry if these questions have already been addressed.)

Kingwl replied:
Thanks for the review.

> What are “duplicated” parameter name hints?

The argument is an identifier and it's same as the parameter name. 
I'm not sure how to accurately describe this behavior. Any suggestions about the naming? 

```
function foo (a: number) {}
var a = 1
foo(a) 
```
> Do we need all three of these options?

[here's](https://github.com/microsoft/vscode/pull/113285#issuecomment-750868023) some context. I'm ok to remove them.

>disabled but the other two enabled?

There's no hints will be provided.

andrewbranch replied:
I actually like the format that Cyrus showed where it was “**suppress** hints when ...” because it makes it more clear that those options won’t do anything if `includeInlayParameterNameHints` is disabled. Also, does VS Code let us show a nested/hierarchical checkbox list like that screenshot? I don’t think so, but that would be really nice.

Kingwl replied:
> suppress hints when ...

I guess that's the description what UI shows. Not about the implementations.
Another problem is VSCode would not open new feature's options by default.
If the option is `suppress something`, the default behavior is `not suppress something`, that is `include/provide something`. (If I'm correct).

> I don’t think so, but that would be really nice.

Sadly it's not supported.

Kingwl replied:
What about:

- includeInlay [xxxxx] HintsWhen [yyyyy]
- include [xxxxx]InlayHintsWhen [yyyyy]

andrewbranch replied:
I think my ideal scenario would be

```ts
includeInlayParameterNameHints?: "none" | "literals" | "all";
suppressInlayParameterNameHintsWhenArgumentMatchesName?: boolean;
```

and the latter can default to `true` both in VS Code and in TS Server. But I’m ok with swapping it to `includeInlayParameterNameHintsWhenArgumentMatchesName` in the options, and maybe trying to use “suppress” language in the VS Code UI and invert the value sent. I just really don’t want the options UI to allow for a bunch of checkboxes that sound like they should enable something but actually do nothing because they’re really sub-options of `includeInlayParameterNameHints`.

andrewbranch reviewed (changes requested):
<img width="503" alt="image" src="https://user-images.githubusercontent.com/3277153/122441017-bd9f2400-cf62-11eb-9f3b-15a809692a40.png">

It looks like we really can’t do return type annotations for arrow functions unless they have parens. I was super confused when I saw this.

Jack-Works commented:
> <img alt="image" width="503" src="https://user-images.githubusercontent.com/3277153/122441017-bd9f2400-cf62-11eb-9f3b-15a809692a40.png">
> 
> It looks like we really can’t do return type annotations for arrow functions unless they have parens. I was super confused when I saw this.

We can add a fake `(` `)` around the unarrowed parans by the inlay hints API

Kingwl commented:
> We can add a fake ( ) around the unarrowed parans by the inlay hints API  

It's a little too noisy

Kingwl commented:
CI failed. But seems not my fault.

andrewbranch commented on src/services/inlayHints.ts:
```diff
@@ -3,6 +3,10 @@ namespace ts.InlayHints {
     const maxHintsLength = 30;
 
+    const leadingParameterNameCommentRegexFactory = (name: string) => {
+        return new RegExp(`^\\s?/\\*\\*?\\s?${name}\\s?\\*\\/\\s?$`);
```
Do we need to sanitize this name before plugging it into a regular expression? There are times when we put non-identifier text into an identifier to help with parser recovery, e.g. leading `#` when you try to use a private identifier where one can’t actually be parsed. I can’t think of a scenario where we put any RegExp control characters in, but maybe it could happen in the future. It would probably be fine just to not run this check if `name` is not valid identifier text.

Kingwl replied:
I guess check the name is a identifierText  would cover this.

Kingwl commented:
Is it possible to take this down in 4.4 beta? Or before RC? (And we could back to vscode side)

xiaoxiangmoe commented:
> Is it possible to take this down in 4.4 beta? Or before RC? (And we could back to vscode side)

It would be great if this feature was available on 4.4！
太好了 早该管管了！

andrewbranch commented:
> Is it possible to take this down in 4.4 beta?

Yes. I’m getting other reviewers to sign off before the end of the week.

Kingwl commented:
Well. Should identifier be a literal argument? Or should we show hints if an argument is an identifier？

jessetrinity commented on tests/cases/fourslash/inlayHintsShouldWork45.ts:
```diff
@@ -0,0 +1,16 @@
+
+goTo.file('/b.js')
+const markers = test.markers();
+verify.getInlayHints([], undefined, {
```
What does this test do?

Kingwl replied:
It's means 'No hints provided'. (as well as the below one)

jessetrinity replied:
Is that meaningful without any markers in the file?

Kingwl replied:
Well... I missed that...😂

jessetrinity commented on tests/cases/fourslash/inlayHintsShouldWork49.ts:
```diff
@@ -0,0 +1,17 @@
+
+goTo.file('/a.js')
+const markers = test.markers();
+verify.getInlayHints([], undefined, {
```
what does this test do?

Kingwl replied:
It's means 'No hints provided'. (as well as the above one)

andrewbranch commented:
> Should identifier be a literal argument?

Well, let me think about some examples:

```ts
f(1);
f(1 + 2);

f("Some String");
f("Some String".toLowerCase());
f(someString);
f(someString.toLowerCase());

f([0, 1, 2]);
f([0, 1, 2].map(processNumber));
f(integers);
f(integers.map(processNumber));
```

In these cases, I feel like the literal arguments and the expressions made up of literals convey the same amount of semantic information. I don’t think the `+ 2` or the `.toLowerCase()` or the `.map(processNumber)` does anything to negate the value of the parameter name hint.

That said, `someString` vs. `someString.toLowerCase()` and `integers` vs. `integers.map(processNumber)` also feel like they convey about the same amount of information. I’m fine just doing the same thing as C#, but the ParenthesizedExpression example in the test looked very strange to me. I think it would be appropriate to `skipParentheses` since we generally recognize that parens have no semantic meaning. But it’s not a big deal. Maybe someone would find it valuable to be able to force a hint to show up by parenthesizing an identifier. It’s not a blocker from me, just curious what the rationale was.

jessetrinity commented on tests/cases/fourslash/inlayHintsShouldWork39.ts:
```diff
@@ -0,0 +1,10 @@
+//// }
+
+const markers = test.markers();
+verify.getInlayHints([], undefined, {
```
no marker in this file either.

jessetrinity commented on tests/cases/fourslash/inlayHintsShouldWork1.ts:
```diff
@@ -0,0 +1,22 @@
+const markers = test.markers();
+verify.getInlayHints([
+    {
+        text: 'a:',
```
Why do all of the hints include a colon (":")?

Kingwl replied:
Not all:
Parameter hints is "name: "
Type hints is ": type"
Enum value hints is "= value"

jessetrinity replied:
ah, so it's meant to sort of "fill out" the syntax at the call site.

Kingwl replied:
Kind of.... I guess

jessetrinity commented on tests/cases/fourslash/inlayHintsShouldWork10.ts:
```diff
@@ -0,0 +1,9 @@
+//// declare const unknownCall: any;
+//// unknownCall();
+
+const markers = test.markers();
```
no markers here either

Kingwl commented:
Thanks for your review. I've updated followed the comments.

Kingwl commented on src/server/protocol.ts:
```diff
@@ -2549,6 +2550,40 @@ namespace ts.server.protocol {
     }
 
+    export const enum InlayHintKind {
+        Other = 0,
```
Yeah. `Other` is defined in vscode side. But I guess we could mapping these two enum on vscode side too.

jessetrinity replied:
Matching vscode names is fine 👍

jessetrinity commented:
I haven't tried this out yet to see how it actually plays. I assume I need vscode insiders?

andrewbranch commented:
@jessetrinity you need a local build of https://github.com/microsoft/vscode/pull/113412

andrewbranch reviewed (approved):
Thanks for bearing with us on the last-minute updates, @Kingwl 🌟

jessetrinity commented on src/services/types.ts:
```diff
@@ -693,6 +705,20 @@ namespace ts {
     }
 
+    export const enum InlayHintKind {
+        Other = 0,
```
Change to match protocol?

jessetrinity commented on tests/baselines/reference/api/tsserverlibrary.d.ts:
```diff
@@ -5841,6 +5851,18 @@ declare namespace ts {
         fromSpans: TextSpan[];
     }
+    enum InlayHintKind {
+        Other = 0,
```
Change to match protocol?

jessetrinity commented on tests/baselines/reference/api/typescript.d.ts:
```diff
@@ -5841,6 +5851,18 @@ declare namespace ts {
         fromSpans: TextSpan[];
     }
+    enum InlayHintKind {
+        Other = 0,
```
Change to match protocol?

jessetrinity commented on tests/cases/fourslash/fourslash.ts:
```diff
@@ -66,6 +66,12 @@ declare module ts {
     }
 
+    const enum InlayHintKind {
+        Other = 0,
```
Change to match protocol?

jessetrinity reviewed (changes requested):
`InlayHintKind` was changed to

```ts
export const enum InlayHintKind {
  Type = "Type",
  Parameter = "Parameter",
  Enum = "Enum",
}
```

in `protocol.ts` but nowhere else.

Kingwl commented:
Thanks!

Avishayy commented:
Thanks for creating such a great feature!

I have a question, why is there a limit on the hint length? Specifically, why is `maxHintsLength` set to 30? Are there any plans to change it sometime?
